### PR TITLE
[Work in Progress] Allow different JAX-RS implementations

### PR DESCRIPTION
**Pull request by chrisruffalo**
This is a preview/WIP of an attempt I made to allow multiple JAX-RS implementations to co-exist with docker-java. For me this is a stumbling block when using docker-cube and EAP because if I am using any Resteasy parts (as part of EAP) then I have a **lot** of classpath manipulation work to do.

What this PR shows is a possible hierarchy that can accommodate Jersy/Resteasy under the same structure.

What works:
- Resteasy and Jersey implementations both test the same (and get the same error with docker 1.8.2)

What is missing:
- SSL Context is broken (it relied on Jersey libraries directly)
- The logging in the SPI parts that load the different implementations
- Unix socket support for Resteasy
- The release/packaging strategy is not defined at all
- Still no JAX-RS 1.0/1.1 support

What could be better:
- The naming, that's probably something I'm least happy with
- Test case layout / structure

All comments welcome! I'm not sure this is even close to the best way because it will require basically all of the commands to be re-implemented (with little reuse) to support earlier versions of JAX-RS and I'm not sure that's the best way.

Another thought was to try and create a CommandTarget (instead of a WebTarget) that has an implementation for JAX-RS 1.0, 1.1, and 2.0. I ran into more issues with that and this approach seemed more straightforward.

### CODE REVIEW

File: api/src/test/java/com/github/dockerjava/api/command/InspectContainerResponseTest.java [hunk 1]
```diff
@@ -33,23 +35,23 @@
 
     @Test
     public void roundTrip_full() throws IOException {
-        InspectContainerResponse[] responses = testRoundTrip(CommandJSONSamples.inspectContainerResponse_full,
+        InspectContainerResponse[] responses = JSONTestHelper.testRoundTrip(CommandJSONSamples.inspectContainerResponse_full,
                 InspectContainerResponse[].class);
-        assertEquals(1, responses.length);
+        Assert.assertEquals(1, responses.length);
```
Comment (lanwen) [inline api/src/test/java/com/github/dockerjava/api/command/InspectContainerResponseTest.java]: Seems like IDE added this full class name after move. If you revert this change (add static import back), diff will be easily to review

File: api/src/test/java/com/github/dockerjava/api/model/BindTest.java [hunk 2]
```diff
@@ -1,9 +1,6 @@
 package com.github.dockerjava.api.model;
 
-import static com.github.dockerjava.api.model.AccessMode.ro;
-import static com.github.dockerjava.api.model.AccessMode.rw;
-import static org.testng.Assert.assertEquals;
-
+import org.testng.Assert;
```
Comment (KostyaSha) [inline api/src/test/java/com/github/dockerjava/api/model/BindTest.java]: Please, don't refactor static imports.

**Comment by chrisruffalo** [comment]
Fixed. I think this came down to IDE "help" and personal preference because I didn't realize how it was being done before.

File: api/src/main/java/com/github/dockerjava/api/DockerClient.java [hunk 3]
```diff
@@ -44,6 +39,11 @@
 import com.github.dockerjava.api.model.AuthConfig;
 import com.github.dockerjava.api.model.Identifier;
 
+import java.io.Closeable;
+import java.io.File;
```
Comment (KostyaSha) [inline api/src/main/java/com/github/dockerjava/api/DockerClient.java]: This also looks like IDE change.

**Comment by chrisruffalo** [comment]
Thanks. Is there a better way to deal with these? Obviously I've "organized" things I shouldn't have.

**Comment by KostyaSha** [comment]
@chrisruffalo seems you need configure your IDE :)

**Comment by chrisruffalo** [comment]
I'm using IntelliJ's defaults... I've not been pleased with them in the past but I like the IDE.

**Comment by marcuslinke** [comment]
@chrisruffalo Thanks for all of your effort, but I don't like the idea of providing different implementations. Although the docker-java API was designed with implementation independency in mind this wasn't meant to provide multiple ones. Rather this was done to eventually switch to ANOTHER implementation later. 

As you already stated, the JAX-RS (jersey) implementation is somewhat problematic because of the existing different JAX-RS runtimes and their incompatibilities. So rather than implementing a second JAX-RS (resteasy) based solution, a better way would be to switch to netty or async-http-client as foundation. This would probably solve most of the problems with compatibility and missing features like redirect local `STDIN` to docker.

**Comment by chrisruffalo** [comment]
@marcuslinke the main issue I see with that is the complexity in creating the REST calls. I've not had much experience doing that without the JAX-RS abstraction.

**Comment by KostyaSha** [comment]
@chrisruffalo you probably may look at https://github.com/kohsuke/github-api that is pretty simple and object oriented. Thought it has no async calls and threadings.

**Comment by chrisruffalo** [comment]
@KostyaSha I'll give it a look. this just isn't a task that's really something that i'm used to working with. I tried doing it this way because I'm comfortable with the scope.

@marcuslinke I forgot to mention: thanks for the clarification on the future and direction.

And, everyone, thanks for the feedback!
